File: 2020/234.py
'''
234
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        # O(n) time and O(1) space?
        # use next : next.next
        if not head: return True
        if not head.next: return True
        if not head.next.next: return head.next.val == head.val

        prev, c1, c2 = head, head.next, head.next.next
        prev.next = None
        # find the middle point while reversing first half
        while c1.next and c2.next and c2.next.next:
            tmp = c1
            c1, c2 = c1.next, c2.next.next
            prev, tmp.next = tmp, prev
        
        # handle if length of ListNode is odd
        if c2.next is None:
            # check from head
            return self.checker(prev, c1.next)
        else:
            r = c1.next
            c1.next = prev
            return self.checker(c1, r)
    
    def checker(self, n1: ListNode, n2: ListNode) -> bool:
        while n1 and n2:
            if n1.val != n2.val:
                return False
            n1, n2 = n1.next, n2.next
        if not n1 and not n2:
            return True
        else:
            return False
            

File: current_session/c++/234.cpp
#include <iostream>
#include <vector>

using namespace std;

// Follow up: Could you do it in O(n) time and O(1) space?
// find the middle point -> reverse the first (or second half) -> compare the LL

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  bool isPalindrome(ListNode* head) {
    ListNode* node = head;
    vector<int> seen;
    while (node != nullptr) {
      seen.push_back(node->val);
      node = node->next;
    }
    int l=0, r=seen.size()-1;
    while (l<r && seen[l] == seen[r]) {
      ++l;
      --r;
    }
    return (l>=r);
  }
};